### ISSUE TITLE
Deprecate /metrics endpoint in favor of Consul discovery

### ISSUE THREAD

**Issue by checketts** [promregator/promregator#158]
## Summary / Problem Statement
The `/metrics` endpoint is a convenient starting point to immediately get metrics from Promregator, since you merely point Prometheus to the Promeregator instance. However it doesn't scale and adds a lot of complexity to Promregator.

### Can't scale
* As more instances are added, the `/metrics` endpoint becomes slower since all instances have to be scraped at the same time. By default Prometheus will timeout the request after 10 seconds. A single slow instance will make all metrics disappear in Prometheus.

### Added complexity
* By scraping several instances and exposing them via a single `/metrics` endpoint Promregator has to add labels to differentiate among the instances called
  * To add labels, Promregator has to parse the existing labels
  * Some labels/metrics can't be created in this way (`up` is one example)
* Promregator has to avoid calling itself to avoid infinite loop.
  * This means Promregator can't be exposed as a typical target, and it has to be treated differently via discovery.
  * By exposing Promregator via a different mechanism, the org/space tags aren't available (see #117)

## Proposed replacement - Consul

The current recommendation is to stop using `/metrics` and use the `/discovery` endpoint. This makes adoption much harder since Prometheus can't scrape the `/discovery` endpoint directly.

A better solution that would minimize Promregator configuration and centralize target configuration/relabeling in Prometheus would be to have Prometheus call Promregator directly for discovery.

This can be done using one of the supported service discovery mechanisms. After investigating several options, the [Consul based one](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#consul_sd_config) is the simplest to implement.

Here is an example of the configuration in Prometheus:
```
scrape_configs:
  - job_name: "promregator"
    scheme: https
    consul_sd_configs:
      - server: https://mypromregator.mydomain.org  #1 
       # services:   #2
       # - my-app
    relabel_configs:
     - source_labels: [__meta_consul_service_metadata_metrics_path] #3 
       action: replace
       target_label: __metrics_path__       
     - source_labels: [__meta_consul_service,__meta_consul_service_metadata_cf_instanceNumber]
       target_label: instance
       separator: "-"
     - source_labels: [__meta_consul_service_metadata_orgName]
       target_label: org_name
     - source_labels: [__meta_consul_service_metadata_spaceName]
       target_label: space_name
     - source_labels: [__meta_consul_service]
       target_label: application
```

- **1** Pointing to a supporting Promregator is all that is needed to begin calling it. Basic auth support is also available
- **2** Services can be filtered in this configuration. Either by relabeling, or listing the 'services' that you want to call
- **3** The `__metrics_path__` is not provided directly by the discovery, so setting it in the relabeling is required. However the value that Promregator discovered is provided via `__meta_consul_service_metadata_metrics_path` and so is available.

While this service discovery is hijacking an existing one that Prometheus supports (consul) and isn't truly running Consul, this could cause confusion. However it will be simple to document, and also support all the features that Consul supports so we stay compatible.

## Benefits

* Simplified setup for new users
* New users will begin using the best and most scalable approach
* Users won't have to maintain a CURL or equivalent mechanism for discovery
* Complexity that exists currently due to the `/metrics` multi-scraping code can be removed.
  * The existing `/discovery` endpoint can be left intact so current users could keep using it and won't have to use the `Consul` facade if they don't want to.
* Promregator `targets` config can be eliminated since all target config can now happen in Prometheus. We can dig into this in a separate issue/PR.

**Comment by eaglerainbow** [comment]
As discussed, this is something for the "major overhaul" of Promregator.

>  and centralize target configuration/relabeling in Prometheus
> [...]
> Promregator targets config can be eliminated since all target config can now happen in Prometheus. We can dig into this in a separate issue/PR.

I didn't get this properly, yet. Why can we get rid of the targets config, if we let Prometheus scrape via consul protocol? Where can we store which orgs/spaces/... need to be scraped using which regular expression in the route in Prometheus config?

**Comment by checketts** [comment]
The Consul API returns a payload like so:

```
[
  {
    "node": "promregator-url.org",
    "datacenter": "cfapi",
    "taggedAddresses": {}, //Not used
    "nodeMeta": {}, //Not used
    "serviceAddress": "promregator-url.org",
    "serviceId": "e287be8c-b6ad-4fbc-af13-ae9f322310a9", //CF App ID
    "serviceName": "myapp-dev", //Application name
    "servicePort": 443, //Promregator port
    "serviceMeta": { // This meta data section is where extra CF data is exposed
      "metrics_path": "/v2/singleTargetScraping/dp-dev/0/-994084376", 
      "cf_api": "ash",
      "cf_orgName": "Marketing",
      "cf_spaceName": "myapp-dev",
      "cf_instanceNumber": "0"
    },
    "serviceTags": [],
    "namespace": "myapp-dev" // CF Space
  }
]
```
Note the `cf_orgName` and similar in the `serviceMeta` field.


Given this structure we can do:
```
  - job_name: "promregator"
    scheme: https
    consul_sd_configs:
      - server: https://promregator-url.org
    relabel_configs:
     - source_labels: [__meta_consul_service]
       regex: 'myapp-.+' 
       action: keep
     - source_labels: [__meta_consul_service_metadata_metrics_path]
       action: replace
       target_label: __metrics_path__
     - source_labels: [__meta_consul_service_metadata_orgName]
       target_label: org_name
     - source_labels: [__meta_consul_service_metadata_spaceName]
       target_label: space_name
     - source_labels: [__meta_consul_service]
       target_label: application
     - source_labels: [__meta_consul_service_metadata_api]
       target_label: api
```

Particular of interest is:
```
     - source_labels: [__meta_consul_service]
       regex: 'myapp-.+' 
       action: keep
```
Where we can use typical Prometheus config to keep/drop targets that are returned from Promregator. This also assumes that Promregator is changed to return all targets if we don't configure one.

One aspect this doesn't address is the `__metrics_path__` that Promregator itself calls, since that wouldn't be easily overridden via Prometheus config. I have some ideas on that regard, but we can save that for a follow up discussion.

**Comment by eaglerainbow** [comment]
> Where we can use typical Prometheus config to keep/drop targets that are returned from Promregator. This also assumes that Promregator is changed to return all targets if we don't configure one.

That was the magic I was missing.
However, that could come at a very high price: I know some folks here that are admins on the platform and hence see all orgs/spaces on the CFCC. However, they might only want to scrape 1% of all these applications out there. In short, for these use cases, we still would require some approach of filtering the list of applications to be discovered. Otherwise, we will chase our poor CFCC with tons of useless metadata requests...

In short, I am totally in line to expose more information via the consul's metadata structure, but I am a little hesitant on getting rid of the target's configuration in Prometheus.

**Comment by eaglerainbow** [comment]
BTW: w.r.t. removal of `/metrics` endpoint: You might be interested at having a look at f4854e092efec7205153bdebffc49df37af52133 , which is available on branch `simplifiedProm`.

**Comment by checketts** [comment]
> In short, I am totally in line to expose more information via the consul's metadata structure, but I am a little hesitant on getting rid of the target's configuration in Prometheus.

Actually that is exactly how I would recommend we proceed:

* Keep the current `targets` config and support for those advanced capabilities
* Seek to minimize the amount of Promregator config that is needed to get is up and running.

**Comment by eaglerainbow** [comment]
Okay, let's see how far we get with your recommended approach.

**Comment by checketts** [comment]
Renaming this issue since consul discovery is no longer needed now that [HTTP discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#http_sd_config) exists as a first class citizen in Prometheus.
